### ISSUE TITLE
[Feature Request]: `WriteLinesToFile` should support specifying a UTF8 encoding that doesn't write a BOM

### ISSUE THREAD

**Issue by baronfel** [dotnet/msbuild#10521]
### Summary

`WriteLinesToFile` accepts an `Encoding` metadata that is a name that's passed to `System.Text.Encoding.GetEncoding`. There is no BCL name for `UTF-8 with no BOM`, so users cannot easily request files be written in this encoding.

### Background and Motivation

BOMs are not generally considered good things to require these days, and many systems (especially those that interact with Linux hosts/tools) have moved away from them because said tools don't handle BOMs well.

We should at least have a supported and documented way to skip BOM emission.

### Proposed Feature

We should support a string like `UTF8-no-bom` or similar for the `Encoding` metadata of the `WriteLinesToFile` Task so that users have an accepted way to write without BOMs. And/or the runtime should provide a named lookup that does the same.

**Comment by newmasterSG** [comment]
@baronfel Hi, can I take this issue?

**Comment by baronfel** [comment]
Yes, that would be great - though we will likely need to coordinate with @terrajobst about what the name for a proper 'UTF8 without BOM' encoding might be for the whole platform to help ensure consistency.

**Comment by newmasterSG** [comment]
So how do I understand that this issue was done?

**Comment by newmasterSG** [comment]
Is this problem solved?

**Comment by terrajobst** [comment]
@baronfel 

> Yes, that would be great - though we will likely need to coordinate with [@terrajobst](https://github.com/terrajobst) about what the name for a proper 'UTF8 without BOM' encoding might be for the whole platform to help ensure consistency.

FWIW, in .NET we default to writing UTF8 without BOM:

```C#
File.WriteAllText("P:\\test.default.txt", message); // Writes UTF8 without BOM
File.WriteAllText("P:\\test.UTF8.txt", message, Encoding.UTF8); // Writes UTF8 with BOM
File.WriteAllText("P:\\test.explicit.txt", message, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false)); Same as default
```

**Comment by newmasterSG** [comment]
May need this issue to close?

**Comment by baronfel** [comment]
I'm uncomfortable with there not being an explicit way to declare UTF8 without BOM' via the Encoding metadata - that's why I wrote the issue in the first place. The presence of a default doesn't mean that we can skip the explicit version of that same default. 

@terrajobst that's why I tagged you here - we need a name for the concept and I didn't want to make names without some kind of discussion, because I could see other encoding-parsing places in the runtime wanting this level of explicitness.

**Comment by terrajobst** [comment]
Both VS Code and Sublime call it

* UTF8 (`utf8`)
* UTF8 with BOM  (`utf8bom`)

Visual Studio calls it (File | Save As | Save with Encoding)

* UTF8 with signature
* UTF8 without signature

The .NET Encoding APIs refer to the BOM as `UTF8Identifier`.

For the MSBuild task I see two options:

* Add another bool property, like `EmitUTF8Identifier`
* Special encoding syntax

If you go with a secondary property, I'd follow the BCL naming conventions. If you go with a special string, I'd probably follow VS Code. Not sure how you feel about which value is the default; if you want `UTF8` to mean "UTF8 with BOM" then I'd probably go with `UTF8_no_bom` or something like this.

/cc @GrabYourPitchforks @tarekgh

**Comment by tarekgh** [comment]
> The .NET Encoding APIs refer to the BOM as UTF8Identifier

That is right. I want to point out encoding support a property called [`Preamble`](https://learn.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.preamble?view=net-9.0). This property was introduced in net core 3.0. So, it looks to me we preferred the name `Preamble`?

> Add another bool property, like EmitUTF8Identifier

I like this option more, but I prefer the name `EmitUTF8Bom` more. I know this is not consistent with previous names, but users recognize `BOM` more than other naming.

**Comment by terrajobst** [comment]
Works for me.

Any thoughts on what the default should be?>

**Comment by tarekgh** [comment]
> Any thoughts on what the default should be?

I prefer to turn off BOM by default. The reason stated above ` many systems (especially those that interact with Linux hosts/tools) have moved away from them because said tools don't handle BOMs well.` which I agree with.
